### ISSUE TITLE
Reviewer feedback on variable safety.

### ISSUE THREAD

**Issue by tobielf** [stable-Kanren/stable-Kanren#3]
The current implementation heavily depends on safe variables, i.e., assigned a value before unification under negation. This restriction makes the program evaluation heavily dependent on the order of conjuncts. For example, two implementations of the `membero` relation produce different results on ground lists: 

```
(defineo (membero x xs)
  (fresh (h t)
      (conso h t xs)
      (conde
        [(== h x)]
        [(membero x t) (noto (== h x))])))

(run 10 (q) (fresh (lst elem) (== lst `(1 2 3)) (== q `(,elem ,lst)) (membero elem lst)))
; ((1 (1 2 3)) (2 (1 2 3)) (3 (1 2 3)))

(run 10 (q) (fresh (lst elem a b c) (== lst `(,a ,b ,c)) (== q `(,elem ,lst)) (membero elem lst)))
; ((_.0 (_.0 _.1 _.2)))
``` 

```
(defineo (membero x xs)
  (fresh (h t)
      (conso h t xs)
      (conde
        [(== h x)]
        [(noto (== h x)) (membero x t)])))

(run 10 (q) (fresh (lst elem) (== lst `(1 2 3)) (== q `(,elem ,lst)) (membero elem lst)))
; ((1 (1 2 3)))

(run 10 (q) (fresh (lst elem a b c) (== lst `(,a ,b ,c)) (== q `(,elem ,lst)) (membero elem lst)))
; ((_.0 (_.0 _.1 _.2)))
```

The first implementation ensures that both `h` and `x` are ground by the time the `noto (== h x)` is executed, while the second implementation does not. It leads to all answers but the first lost. What is worse, this happens silently, which is bound to complicate debugging a lot. Both implementations fail to find more than a single answer when the list is not completely ground.

**Comment by tobielf** [comment]
The bug seems tricky at first glance; the first definition of `membero` is valid (variables are safe), but our solver still produced the answer missing `(_.0 (_.1 _.0 _.2))` and `(_.0 (_.1 _.2 _.0))`.

After inspecting the runtime information, it turned out to be the BFS behavior of the interleaved `conde` causing the problem. We are expecting `(membero x t)` and `(conso h t x)` to unify `x` and `h` with a value before `(noto (== x h))`. However, the interleaving behavior delayed value assignment, so both `x` and `h` are free variables at the time of `(noto (== x h))`

One solution is to use [`conde-dfs`](https://github.com/webyrd/Barliman/blob/adf8b3146737f4682dd73fab12f4f711e93d508a/cocoa/Barliman/mk-and-rel-interp/mk/mk.scm#L447) to guarantee the variables are safe before using it under negation (for this case). But we still have to deal with other free variable issues under DFS.

The other solution is to add a constraint system to add an inequality constraint to two free variables. The solver keeps running (conditional success) until these free variables assignment violates the constraint later. It's like adding a watcher to a variable and allowing the use of free variables under the negation goal, whether under DFS or BFS. So, we are adapting this approach.

**Comment by tobielf** [comment]
I think having a set of debugging utilities (framework) like `(inspecto var1 var2)` would be good for printing out useful information at runtime. Also, toggling the version of `conde` between BFS and DFS. Debugging BFS information is like a multithread stacktrace. DFS would be easier.

**Comment by tobielf** [comment]
Merged #4
